### ISSUE TITLE
Terraform state pull to do less syntax checking

### ISSUE THREAD

**Issue by gtmtech** [hashicorp/terraform#32598]
### Terraform Version

```shell
Terraform v1.2.3
```

### Use Cases

I want to get the terraform state file quickly and then process it offline. I need to do this because I need to do some operations on it that terraform itself is not good at. Moving things around a bit, converting from lists of count=1 back to singleton objects and things like that. It's a lot quicker if I can pull the statefile and process it and push it back. 

terraform state pull however makes lots of complaints about this or that module not being installed - ie it is trying to do a reconciliation of the code against the statefile. Precisely the problem I am trying to fix by pulling the statefile and modifying and pushing it back. 

```
│ Error: Module not installed
│ 
│   on x/y/z.tf line 42:
│   42: module "foo" {
│ 
│ This module is not yet installed. Run "terraform init" to install all
│ modules required by this configuration.
```

Pulling a state as a conceptual thing has no dependencies on whether the code actually works, other than it has to understand where the code says the backend configuration is - and thats it. 

It would be useful if the commands do the minimum they need in terms of their dependencies.

### Attempted Solutions

In the end I had to copy the state out manually using `aws s3` command - but I consider this to be a usecase for which the `terraform state pull` was not as helpful as it needed to be.

**Comment by apparentlymart** [comment]
Hi @gtmtech! Thanks for sharing this use-case.

I've not looked closely at the code for this command for some time now but at first glance it seems like the part that is causing this error is the step to check whether the current Terraform version is appropriate for the current configuration, which involves checking for the `required_version` argument in all modules:

https://github.com/hashicorp/terraform/blob/fc8fed0047e98b430f648c3659d22be6f8d52642/internal/command/state_pull.go#L26-L29

That seems to have originated over in https://github.com/hashicorp/terraform/pull/30511, where there was some discussion about the tradeoff of safety vs. flexibility, and was intended to deal with https://github.com/hashicorp/terraform/issues/28025 which was to avoid accidentally generating new state snapshots with a newer version of Terraform that might then be incompatible with the Terraform version that the configuration is intended for use with.

Based on that discussion it seems like the intent here was to limit this only to commands which can generate new state snapshots; it seems that still including `terraform state pull` in there was an oversight as I would've expected that to also be classified as a read-only `state` subcommand command.

It seems reasonable to me then to remove this check from `terraform state pull`, but first I'd like to hear if @alisdair remembers this discussion and whether there was a strong reason to include the "pull" command under this that I'm not thinking of.

Thanks again!

**Comment by alisdair** [comment]
Nope, that was an oversight on my part. I think it would be fine to remove this check from `state pull`.

**Comment by jbardin** [comment]
One thing to consider is that `state pull` re-encodes the state at the version of the CLI, so if the state were say written by 0.13, and the config required that version, checking the config first would prevent that unintended upgrade.

**Comment by apparentlymart** [comment]
My assumption would be that we'd check on the way back in using `terraform state push`, and so having the local file in a newer format isn't dangerous.

However, despite not being dangerous it is still potentially confusing and annoying. In particular, it would be annoying to have spent time manually editing a state snapshot (assuming that's the goal) and then find that Terraform refuses to upload that snapshot.
